skimage/morphology/grey.py

```diff
@@ -1,17 +1,133 @@
 """
 Grayscale morphological operations
 """
-from skimage import img_as_ubyte
-from .misc import default_fallback
-
-from . import cmorph
+import functools
 
+import numpy as np
+from scipy import ndimage as nd
+from .misc import default_selem
+from ..util import pad, crop
 
 __all__ = ['erosion', 'dilation', 'opening', 'closing', 'white_tophat',
            'black_tophat']
 
 
-@default_fallback
+def _shift_selem(selem, shift_x, shift_y):
+    """Shift the binary image `selem` in the left and/or up.
+
+    This only affects structuring elements with even number of rows or
+    columns.
+
+    Parameters
+    ----------
+    selem : 2D array, shape (M, N)
+        The input structuring element.
+    shift_x, shift_y : bool
+        Whether to move `selem` along each axis.
+
+    Returns
+    -------
+    out : 2D array, shape (M + int(shift_x), N + int(shift_y))
+        The shifted structuring element.
+    """
+    if selem.ndim > 2:
+        # do nothing for 3D or higher structuring elements
+        return selem
+    m, n = selem.shape
+    if m % 2 == 0:
+        extra_row = np.zeros((1, n), selem.dtype)
+        if shift_x:
+            selem = np.vstack((selem, extra_row))
+        else:
+            selem = np.vstack((extra_row, selem))
+        m += 1
+    if n % 2 == 0:
+        extra_col = np.zeros((m, 1), selem.dtype)
+        if shift_y:
+            selem = np.hstack((selem, extra_col))
+        else:
+            selem = np.hstack((extra_col, selem))
+    return selem
+
+
+def _invert_selem(selem):
+    """Change the order of the values in `selem`.
+
+    This is a patch for the *weird* footprint inversion in
+    `nd.grey_morphology` [1].
+
+    Parameters
+    ----------
+    selem : array
+        The input structuring element.
+
+    Returns
+    -------
+    inverted : array
+        The structuring element, in opposite order.
+
+    Examples
+    --------
+    >>> selem = np.array([[0, 0, 0], [0, 1, 1], [0, 1, 1]], np.uint8)
+    >>> _invert_selem(selem)
+    array([[1, 1, 0],
+           [1, 1, 0],
+           [0, 0, 0]], dtype=uint8)
+
+    References
+    ----------
+    [1] https://github.com/scipy/scipy/blob/ec20ababa400e39ac3ffc9148c01ef86d5349332/scipy/ndimage/morphology.py#L1285
+    """
+    inverted = selem[(slice(None, None, -1),) * selem.ndim]
+    return inverted
+
+
+def pad_for_eccentric_selems(func):
+    """Pad input images for certain morphological operations.
+
+    Parameters
+    ----------
+    func : callable
+        A morphological function, either opening or closing, that
+        supports eccentric structuring elements. The inputs must
+        include at least `image`, `selem`, and `out`.
+
+    Returns
+    -------
+    func_out : callable
+        A function
+
+    See Also
+    --------
+    ``opening``, ``closing``.
```

Comment:
Neither am I. =P What do you mean? These are function names. Are they not in backticks normally?